docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,462 @@
+====================
+Expressions in Relay
+====================
+
+The Relay IR is a pure, expression-oriented language with distinct
+dataflow and control flow language fragments.
+Each dataflow fragment of a program (i.e., the portions of the
+program without recursive calls or branching) can be viewed as a
+traditional computation graph when writing and expressing transformations.
+
+The below sections describe the different expressions in Relay
+as well as give details of their semantics.
+
+Variables
+=========
+
+Inspired by LLVM, Relay explicitly distinguishes between local and
+global variables both in the AST and in the text format. In the text format,
+global and local variables are distinguished by prefixes, or sigils.
+Global variables are prefixed with :code:`@` and local variables with :code:`%`.
+
+This explicit distinction makes certain optimizations easier to implement.
+For example, inlining a global definition requires no analysis: simply
+substituting the definition suffices.
+
+Global Variable
+~~~~~~~~~~~~~~~~~~
+
+Global identifiers are prefixed by the :code:`@` sigil, such as ":code:`@global`".
+A global identifier always references a globally visible definition contained in the environment.
+Global identifiers must be unique.
+
+See :py:class:`~tvm.relay.expr.GlobalVar` for its implementation
+and documentation.
+
+Local Variable
+~~~~~~~~~~~~~~
+
+Local identifiers are prefixed by the :code:`%` sigil,
+such as ":code:`%local`". A local identifier always references
+a function argument or a variable bound in a :code:`let` expression,
+and will be scoped to the function where it appears or the :code:`let`
+expression where it is bound, respectively.
+
+In the below code segment, notice that :code:`%a` is defined twice. This is
+permitted, as in most functional languages; in the scope of the second
+:code:`let` expression, the name :code:`%a` is "shadowed," meaning all
+references to :code:`%a` in the inner scope refer to the later defintion, while
+references to :code:`%a` in the outer scope continue to refer to
+the first one.
+
+.. code-block:: python
+  def @f(%a) {
+    let %b = %a;
+    let %a = add(%a, %a);
+    multiply(%a, %b)
+  }
+
+(Note that in Relay's implementation, each definition of a local variable
+creates a new :py:class:`~tvm.relay.expr.Var`, so a shadowed local variable,
+despite having the same name as one in an outer scope, will be a different
+object. This allows for comparing local variables by pointer identity with the
+knowledge that the same local variable object corresponds to a different binding site.)
+
+See :py:class:`~tvm.relay.expr.Var` for its implementation
+and documentation.
+
+Global Functions
```

Comment:
structural comment: Why do we have a separate section for Global Functions when the AST has no such explicit construct? I understand why we need to discuss the text format, since the syntax is different, but much of this explanation applies more specifically to the Function node than to global functions. Type relations perhaps need to remain in this section, too.

Reply:
Good question. It was in the first draft so I kept it as it was. A possible reason to keep it as a separate section is that a Relay program can be thought of as a collection of global functions. Not too much of it is specific to global functions, you're right; I will make sure the remark about local functions doesn't repeat anything from this section.